gridcomps/ExtData3G/DataSetNode.F90

```diff
@@ -0,0 +1,285 @@
+   end function equals
+
+   subroutine reset(this)
+      class(DataSetNode), intent(inout) :: this
+      deallocate(this%file)
+      this%enabled = .false.
+      this%update = .false.
+   end subroutine
+
+   function validate(this, current_time, rc) result(node_is_valid)
+      logical :: node_is_valid
+      class(DataSetNode), intent(inout) :: this
+      type(ESMF_Time), intent(in) :: current_time
+      integer, intent(out), optional :: rc
+
+      integer :: status
+      if (.not.allocated(this%file)) then
+         node_is_valid = .false.
+         _RETURN(_SUCCESS)
+      end if
+      if (this%node_side == NODE_UNKNOWN ) then
+         node_is_valid = .false.
+         _RETURN(_SUCCESS)
+      end if
+      if (this%node_side == NODE_LEFT) then
+         node_is_valid = (current_time >= this%file_time)
+      else if (this%node_side == NODE_RIGHT) then
+         node_is_valid = (current_time < this%file_time)
+      end if
+      _RETURN(_SUCCESS)
+   end function
+
+   subroutine invalidate(this)
+      class(DataSetNode), intent(inout) :: this
+      if (allocated(this%file)) then
+         deallocate(this%file) 
+      end if
+      this%enabled = .false.
+      this%update = .false.
+   end subroutine
+
+   subroutine update_node_from_file(this, filename, target_time, rc)
+      class(DataSetNode), intent(inout) :: this
+      character(len=*), intent(in) :: filename
+      type(ESMF_Time), intent(in) :: target_time
+      integer, optional, intent(out) :: rc
+
+      integer :: status, i
+      type(FileMetaDataUtils) :: metadata
+      type(FileMetadata) :: basic_metadata
+      type(NetCDF4_FileFormatter) :: formatter
+      type(ESMF_Time), allocatable :: time_vector(:)
+
+      _ASSERT(this%node_side/=NODE_UNKNOWN, "node does not have a side")
+      call formatter%open(filename, pFIO_READ, _RC)
+      basic_metadata = formatter%read(_RC)
+      call formatter%close()
+      metadata = FileMetadataUtils(basic_metadata, filename)
+
+      call metadata%get_time_info(timeVector=time_vector, _RC)
+      select case(this%node_side)
+      case (NODE_LEFT)
+         do i=size(time_vector),1,-1
+            if (target_time >= time_vector(i)) then
+               this%file = filename
+               this%interp_time = time_vector(i)
+               this%file_time = time_vector(i)
+               this%time_index = i
+               this%enabled = .true.
+               this%update = .true.
+               exit
+            end if
+         enddo
+      case (NODE_RIGHT)
+         do i=1,size(time_vector)
+            if (target_time < time_vector(i)) then
+               this%file = filename
+               this%interp_time = time_vector(i)
+               this%file_time = time_vector(i)
+               this%time_index = i
+               this%enabled = .true.
+               this%update = .true.
+               exit
+            end if
+         enddo
+      end select
```

Comment:
These blocks are identical except for loop order.   I suggest using the SELECT CASE just to determine the triplet: {i0, i1, sign}  and then have a single loop.   You'll thank me when you need to add another component to the data structure.

Reply:
OK - I'll contemplate whether something along these lines could still be worthwhile.  Leave this unresolved for now.